### PR TITLE
Fix dynamo issue

### PR DESCRIPTION
**Pull request by oraluben**
Dynamo use faketensor to trace tensor ops. In some case, the mechanism break compiling with deepspeed.

An example could be found at https://gist.github.com/oraluben/9b8240c2fe482eb4382453d6c97a5f76, to see issues, install deepspeed==0.14.4 instead of my fork

without this PR, llama cannot be compiled.

Detailed explanation:

1. `ZeROOrderedDict`
    dynamo use deepcopy to copy tensors, which will call `object.__reduce__`. When copying `ZeROOrderedDict`, the default implementation do not copy its `_parent_module` and will lead to failure.
2. `param` maybe faketensor and do not have `ds_status` yet, but during tracing it's ok to just skip the `register_external_parameter`, it should be done ways before.

### CODE REVIEW

**Review by tohtana — APPROVED** [review]
@oraluben Thank you for offering a great investigation! I think this is a clean and simple solution for the issue.

File: deepspeed/runtime/zero/parameter_offload.py [hunk 1]
```diff
@@ -49,14 +49,18 @@ def __init__(self, parent_module, *args, **kwargs):
         self._parent_module = parent_module
         self._in_forward = False
 
+    def __reduce__(self):
+        r0, _, *r2 = super().__reduce__()
+        return r0, (self._parent_module, ), *r2
```
Comment (loadams) [inline deepspeed/runtime/zero/parameter_offload.py:54]: @oraluben - it looks like this syntax isn't supported on python 3.7, is this something you could try to generalize for that as well?

Comment (oraluben) [reply deepspeed/runtime/zero/parameter_offload.py:54]: That's a surprise 😂 working on it

**Comment by oraluben** [comment]
`torch.compiler.is_compiling()` should be better for this case, however there's still issue, presumably on dynamo side. So keep it for now.

```
[rank1]:   File "/home/yyc/accelerate-demo/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1720, in __getattr__
[rank1]:     return _parameters[name]
[rank1]:   File "/home/yyc/repo/DeepSpeed/deepspeed/runtime/zero/parameter_offload.py", line 67, in __getitem__
[rank1]:     if not is_compiling() and param.ds_status == ZeroParamStatus.NOT_AVAILABLE:
[rank1]: torch._dynamo.exc.TorchRuntimeError: Failed running call_module L__self___self_attn_q_proj(*(FakeTensor(..., device='cuda:1', size=(1, s0, 4096), dtype=torch.float16,
[rank1]:            grad_fn=<MulBackward0>),), **{}):
[rank1]: 'FakeTensor' object has no attribute 'ds_status'
```

**Comment by loadams** [comment]
@oraluben - sorry this PR has taken so long to be merged, I think it just needed to have master merged again to get the XPU fixes.

**Review by loadams — APPROVED** [review]

**Pull request merged**